Refactor key derivation structure for the driver interface + Implement key derivation input getters

mprse opened the pull request:
## Description
- Refactor the key derivation data structures (struct psa_key_derivation_s and subtructures) to separate the data used by phase 1 (driver-independent) and the data used by phase 2 (specific to our implementation). Call the structure used for phase 1 psa_crypto_driver_key_derivation_inputs_t.
- Refactor the existing code accordingly, with no expected behavior change.
- Implement key derivation input getters
- Create unit tests for these functions
- Use these functions instead of direct access in `psa_crypto_key_derivation_complete_inputs`

Resolves:
https://github.com/Mbed-TLS/mbedtls/issues/5477
https://github.com/Mbed-TLS/mbedtls/issues/5478

This PR needs https://github.com/Mbed-TLS/mbedtls/pull/5834, so the changes are made on top of this #5834.

## Status
**IN PROGRESS**

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -4388,11 +4388,67 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
-        ( ! hkdf->info_set && ! PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) ) )
-        return( PSA_ERROR_BAD_STATE );
-    hkdf->state = HKDF_STATE_OUTPUT;
+    if ( hkdf->inputs.state != HKDF_STATE_OUTPUT )
```
This whole block is part of the input gathering stage, and should be in its own function. Furthermore, I think it's overly complex to call this from an algorithm-specific function: the code to finish processing the inputs and go from the input stage to the output stage should be called directly from `output_bytes`. The logic in `output_bytes` is:

1. If the operation is still at the input stage, call `psa_crypto_key_derivation_complete_inputs` (new function name).
2. Check the capacity.
3. Dispatch to the appropriate `read` function.

`psa_crypto_key_derivation_complete_inputs` checks that all required inputs are present and copies data from the `inputs` side of the union to the `running_context` part. Later it will dispatch to a driver if there is one.

mprse replied:
Good point. I will move this part to separate function (`psa_crypto_key_derivation_complete_inputs`) that will be called from `output_bytes()` to finish gathering stage.

At least in the current proposition `inputs` are already part of the `running_context` and initialized so no copy operation is required.

mprse commented on include/psa/crypto_struct.h:
```diff
@@ -245,20 +261,40 @@ typedef struct psa_tls12_prf_key_derivation_s
     unsigned int MBEDTLS_PRIVATE(can_output_key) : 1;
     size_t MBEDTLS_PRIVATE(capacity);
+
     union
```
> The context is now organized by stage first, and by implementation second, instead of being organized by algorithm first  and more or less by stage inside each algorithm

Currently (`development`) the `context` is organized by algorithm(hkdf vs prf) first:

```
    union
    {
        /* Make the union non-empty even with no supported algorithms. */
        uint8_t MBEDTLS_PRIVATE(dummy);
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
        psa_hkdf_key_derivation_t MBEDTLS_PRIVATE(hkdf);
#endif
#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF) || \
    defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS)
        psa_tls12_prf_key_derivation_t MBEDTLS_PRIVATE(tls12_prf);
#endif
    } MBEDTLS_PRIVATE(ctx);
```

What do you exactly mean by: `stage`, `implementation`, `algorithm`?

gilles-peskine-arm replied:
Stage: 1 = gathering inputs, 2 = producing output

Algorithm: HKDF, TLS12_PRF, ...

Implementation: currently one for each algorithm, but the next step after this task is to support drivers

mprse commented:
Pushed the next version with the following changes:
- `psa_key_derivation_s`: contains union of `inputs/context`. 
- `psa_key_derivation_s`: added `stage` to distinguish inputs stage from implementation specific context stage.
- Refactored code as suggested. Added `psa_crypto_key_derivation_complete_inputs` called from `output_bytes` that  checks that all required inputs are present and perform stage transition (switch from input stage to specific context).

Now we don't have `psa_crypto_key_derivation_driver_setup` as `psa_crypto_key_derivation_complete_inputs` does the stage transition.

Tests are passing.

I was thinking do we need `state`, (`info_set`) ? If the stages are separated then inputs can be provided in any order, but then we would need to adapt the tests also.

mprse commented:
Added getter functions (https://github.com/Mbed-TLS/mbedtls/issues/5478) and adapted only `psa_crypto_key_derivation_complete_hkdf_inputs` for now. 
I followed the requirements and copied inputs from one buffer to another (`inputs`->`impl context`), but for me this is redundant as we could only set pointers/sizes in `context` structure.

mprse commented:
Unfortunately rebase is now required, but need to wait for https://github.com/Mbed-TLS/mbedtls/pull/5834 first.

mprse commented:
Closed per https://github.com/Mbed-TLS/mbedtls/pull/5967/.
